File: Python/Baekjoon/[Gold V] 최소비용 구하기 (1916번).py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


import heapq
import sys
input = sys.stdin.readline

def dijkstra(start):
    q = []
    heapq.heappush(q,(0,start))
    distance[start] = 0
    
    while q:
        dist, node = heapq.heappop(q)
        
        if dist>distance[node]:
            continue
        for i in graph[node]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q,(cost, i[0]))


N = int(input())
M = int(input())
INF = 999999999

distance = [INF]*(N+1)
graph = [[] for _ in range(N+1)]

for _ in range(M): 
    u, v, w = map(int, input().split())
    graph[u].append((v,w))
    
s, e = map(int, input().split())

dijkstra(s)

print(distance[e]-distance[s])

File: Java/Baekjoon/[Gold V] 최소비용 구하기 (1916번).java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Edge implements Comparable<Edge>{
    int u, v, w;

    public Edge(int u, int v, int w){
        this.u = u;
        this.v = v;
        this.w = w;
    }

    @Override
    public int compareTo(Edge other){
        if (this.w != other.w){
            return Integer.compare(this.w, other.w);
        }
        else if (this.u != other.u){
            return Integer.compare(this.u, other.u);
        }
        else{
            return Integer.compare(this.v, other.v);
        }
    }
}

class Main{
    static int[] distance;
    static ArrayList<ArrayList<Edge>> graph;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        StringTokenizer st;
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        distance = new int[n+1];
        for (int i = 0; i<n+1; i++){
            distance[i] = 99999999;
        }

        graph = new ArrayList<>();
        for (int i = 0; i<n+1; i++){
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i<m; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            graph.get(a).add(new Edge(a, b, w));
        }
        st = new StringTokenizer(br.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        dijkstra(start);

        System.out.println(distance[end]-distance[start]); 
    }    
    public static void dijkstra(int start){
        PriorityQueue<Edge> q = new PriorityQueue<>();
        q.add(new Edge(start, start, 0));
        distance[start] = 0;

        while (!q.isEmpty()){
            Edge ed = q.poll();
            int dist = ed.w;
            int node = ed.v;

            if (dist>distance[node]){
                continue;
            }
            for (int i = 0; i<graph.get(node).size() ;i++){
                int cost = dist + graph.get(node).get(i).w;
                if (cost < distance[graph.get(node).get(i).v]){
                    distance[graph.get(node).get(i).v] = cost;
                    q.add(new Edge(graph.get(node).get(i).u,graph.get(node).get(i).v, cost));
                }
            }
        }
    }
}